ghcide/src/Development/IDE/Core/Compile.hs

```diff
@@ -532,45 +532,47 @@ indexHieFile se mod_summary srcPath hash hf = atomically $ do
     -- Get a progress token to report progress and update it for the current file
     pre = do
       tok <- modifyVar indexProgressToken $ \case
-        x@(Just tok) -> pure (x, tok)
+        x@(Just _) -> pure (x, x)
         -- Create a token if we don't already have one
         Nothing -> do
-          u <- LSP.ProgressTextToken . T.pack . show . hashUnique <$> newUnique
-          lspId <- getLspId se
-          eventer se $ LSP.ReqWorkDoneProgressCreate $
-            LSP.fmServerWorkDoneProgressCreateRequest lspId $
-              LSP.WorkDoneProgressCreateParams { _token = u }
-          eventer se $ LSP.NotWorkDoneProgressBegin $
-            LSP.fmServerWorkDoneProgressBeginNotification
-              LSP.ProgressParams
-                  { _token = u
-                  , _value = LSP.WorkDoneProgressBeginParams
-                    { _title = "Indexing references from:"
-                    , _cancellable = Nothing
-                    , _message = Nothing
-                    , _percentage = Nothing
-                    }
-                  }
-          pure (Just u, u)
+          case lspEnv se of
+            Nothing -> pure (Nothing, Nothing)
+            Just env -> LSP.runLspT env $ do
+              u <- LSP.ProgressTextToken . T.pack . show . hashUnique <$> liftIO newUnique
+              b <- liftIO newBarrier
+              _ <- LSP.sendRequest LSP.SWindowWorkDoneProgressCreate (LSP.WorkDoneProgressCreateParams u) (liftIO . signalBarrier b)
+              -- Wait for the progress create response to use the token
+              resp <- liftIO $ waitBarrier b
+              case resp of
+                -- We didn't get a token from the server
+                Left _err -> pure (Nothing,Nothing)
+                Right _ -> do
+                  LSP.sendNotification LSP.SProgress $ LSP.ProgressParams u $
+                    LSP.Begin $ LSP.WorkDoneProgressBeginParams
+                      { _title = "Indexing references from:"
+                      , _cancellable = Nothing
+                      , _message = Nothing
+                      , _percentage = Nothing
+                      }
+                  pure (Just u, Just u)
```

Comment:
It looks like all the branches uses `pure (x, x)`. How about extracting `dupe` from these to make it clear that we are returning the value of `Var`?